### PR TITLE
[API] Prototype of C API & test for Neural Network Model

### PR DESCRIPTION
**Pull request by jijoongmoon**
- Add C API to construct / destruct Neural Network Model object
. ml_nnmodel_construct()
. ml_nnmodel_destruct()

- Add test cases

**Self evaluation:**
1. Build test:	 [X]Passed [ ]Failed [ ]Skipped
2. Run test:	 [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: jijoong.moon <jijoong.moon@samsung.com>

### CODE REVIEW

File: api/capi/src/nntrainer.cpp [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+/**
+ * @brief Function to create Network::NeuralNetwork object.
+ */
+static void *nn_object(void) { return (void *)(new Network::NeuralNetwork()); }
+
+int ml_nnmodel_construct(ml_nnmodel_h *model) {
+  int status = ML_ERROR_NONE;
+  *model = nn_object();
+  if (model == NULL)
```
[the remark this answers is not in the document]
Comment (jijoongmoon) [reply api/capi/src/nntrainer.cpp:40]: As you can see, nn_object() return the class object pointer. This is for the check if it is properly assigned to model. and if it is null, that means there is something wrong with assign object.

File: api/capi/include/nntrainer.h [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+} ml_error_e;
+
+/**
+ * @brief Constructs the neural network model.
+ * @details Use this function to create Neural Netowrk Model.
+ * @since_tizen 5.5
+ * @param[in] model The NNTrainer Model handler from the given description.
+ * @return @c 0 on success. Otherwise a negative error value.
+ * @retval #ML_ERROR_NONE Successful
+ * @retval #ML_ERROR_CANNOT_ASSIGN_ADDRESS Cannot assign object.
```
Comment (jaeyun-jung) [inline api/capi/include/nntrainer.h:69]: recommended to run checker before the ACR process.
Please see internal guideline for native API.
```
nntrainer.h:71: Question:
  Are you sure this function couldn't be given invalid parameter values? Please consider adding INVALID_PARAMETER error.

nntrainer.h:74: Question:
  Are you sure @brief is in third person (Gets ..., Sets ...)?
```

Comment (jijoongmoon) [reply api/capi/include/nntrainer.h:69]: Thanks. It will help me a lot.

File: api/capi/include/nntrainer.h [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+ */
+typedef enum {
+  ML_ERROR_NONE = TIZEN_ERROR_NONE,                                   /**< Success! */
+  ML_ERROR_INVALID_PARAMETER = TIZEN_ERROR_INVALID_PARAMETER,         /**< Invalid parameter */
+  ML_ERROR_UNKNOWN = TIZEN_ERROR_UNKNOWN,                             /**< Unknown error */
+  ML_ERROR_TIMED_OUT = TIZEN_ERROR_TIMED_OUT,                         /**< Time out */
+  ML_ERROR_NOT_SUPPORTED = TIZEN_ERROR_NOT_SUPPORTED,                 /**< The feature is not supported */
+  ML_ERROR_PERMISSION_DENIED = TIZEN_ERROR_PERMISSION_DENIED,         /**< Permission denied */
+  ML_ERROR_OUT_OF_MEMORY = TIZEN_ERROR_OUT_OF_MEMORY,                 /**< Out of memory (Since 6.0) */
+  ML_ERROR_CANNOT_ASSIGN_ADDRESS = TIZEN_ERROR_CANNOT_ASSIGN_ADDRESS, /**< CAnnot assign requested address */
```
Comment (jaeyun-jung) [inline api/capi/include/nntrainer.h:58]: typo : CAnnot

Comment (jijoongmoon) [reply api/capi/include/nntrainer.h]: Ooops. Thanks.

File: api/capi/src/nntrainer.cpp [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+  int status = ML_ERROR_NONE;
+  ml_nnmodel *nnmodel = new ml_nnmodel;
+  nnmodel->magic = ML_NNTRAINER_MAGIC;
+
+  *model = nnmodel;
+
+  try {
+    nnmodel->network = new Network::NeuralNetwork();
+  } catch (const char *e) {
+    std::cerr << "Error: heep exception: " << e << std::endl;
```
Comment (zhoonit) [inline api/capi/src/nntrainer.cpp:65]: I think this is typo `heep` -> `heap`

Comment (jijoongmoon) [reply api/capi/src/nntrainer.cpp:65]: Ooops. Thanks.

File: jni/Android.mk [hunk 5]
```diff
@@ -34,7 +34,7 @@ INIPARSER_SRCS := $(INIPARSER_ROOT)/src/iniparser.c \
 INIPARSER_INCLUDES := $(INIPARSER_ROOT)/src
 
 LOCAL_ARM_NEON      := true
-LOCAL_CFLAGS        += -pthread -fopenmp
+LOCAL_CFLAGS        += -pthread -fopenmp -fexceptions
```
Comment (jaeyun-jung) [inline jni/Android.mk:37]: consider to add -frtti
https://developer.android.com/ndk/guides/cpp-support#rtti

File: api/capi/src/nntrainer.cpp [hunk 6]
```diff
@@ -0,0 +1,101 @@
+/**
+ * Copyright (C) 2019 Samsung Electronics Co., Ltd. All Rights Reserved.
```
Comment (jaeyun-jung) [inline api/capi/src/nntrainer.cpp:2]: plz consider to change 2019 > 2020 (all added files)

TODO:
log util (dlog for tizen)

overall lgtm

Comment (jijoongmoon) [reply api/capi/src/nntrainer.cpp]: I will implement log later. Thanks.

File: jni/Android.mk (hunk 5 above)
Comment (jijoongmoon) [reply jni/Android.mk:37]: added!

**Pull request merged**